include/engine/engine.hpp

```diff
@@ -30,38 +32,162 @@ namespace osrm
 namespace engine
 {
 
-class Engine final
+class EngineInterface
+{
+  public:
+    virtual ~EngineInterface(){};
+    virtual Status Route(const api::RouteParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Table(const api::TableParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Nearest(const api::NearestParameters &parameters,
+                           util::json::Object &result) const = 0;
+    virtual Status Trip(const api::TripParameters &parameters,
+                        util::json::Object &result) const = 0;
+    virtual Status Match(const api::MatchParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Tile(const api::TileParameters &parameters, std::string &result) const = 0;
+};
+
+template <typename AlgorithmT> class Engine final : public EngineInterface
 {
   public:
-    explicit Engine(const EngineConfig &config);
+    explicit Engine(const EngineConfig &config)
+        : route_plugin(config.max_locations_viaroute),       //
+          table_plugin(config.max_locations_distance_table), //
+          nearest_plugin(config.max_results_nearest),        //
+          trip_plugin(config.max_locations_trip),            //
+          match_plugin(config.max_locations_map_matching),   //
+          tile_plugin()                                      //
+
+    {
+        if (config.use_shared_memory)
+        {
+            facade_provider = std::make_unique<WatchingProvider<AlgorithmT>>();
+        }
+        else
+        {
+            facade_provider =
+                std::make_unique<ImmutableProvider<AlgorithmT>>(config.storage_config);
+        }
+    }
 
     Engine(Engine &&) noexcept = delete;
     Engine &operator=(Engine &&) noexcept = delete;
 
     Engine(const Engine &) = delete;
     Engine &operator=(const Engine &) = delete;
+    virtual ~Engine(){};
+
+    Status Route(const api::RouteParameters &params,
```

Comment:
Hrm any good ideas for an alternative? I tried to come up with something better but it would either break the abstraction of `RoutingAlgorithms` (like a `GetFacade()` function) or just as ugly:
```
std::shared_ptr<datafacade::ContiguousDataFacade> facade;
RoutingAlgorithms algos;
std::tie(facade, algos) = GetInstance();
pluging->HandleRequest(facade, algos, ...);
```